### PR TITLE
Spectator-go Thin Client re-write

### PR DESCRIPTION
**Pull request by jivimberg**
We are migrating [spectator-go](https://github.com/Netflix/spectator-go) to a thin-client implementation that uses [spectatord](https://github.com/Netflix-Skunkworks/spectatord). 

The goal is to make  it more maintainable by reducing the API surface area and to bring consistency across platforms (namely: [spectator-reg-sidecar](https://github.com/Netflix/spectator/tree/main/spectator-reg-sidecar/), [spectator-py](https://github.com/Netflix/spectator-py), [spectator-js](https://github.com/Netflix/spectator-js))

As part of this re-write we decided to:

1. Make meters stateless.
2. Introduce a customizable writers (noop, memory, file, UDP, etc).
3. By default the Registry will use a UDP writer configured to hit spectatord sidecar.
4. Removed unused methods to simplify API and stay consistent across platforms.
5. Simplified config options.
6. Moved runtime metrics functionality to [spectator-go-runtime-metrics](https://github.com/Netflix/spectator-go-runtime-metrics).
7. Removed HttpClient because it's no longer required.
8. Added a set of default common tags read from environment variables.

### CODE REVIEW

**Review by copperlight — CHANGES_REQUESTED** [review]
* We should add a `GaugeWithTTL` method, or a means to specify a TTL on the existing Gauge methods. This is an occasional request and a nice feature to provide.
* We should update the README with migration instructions from the thick client version to the thin client version.
* We should provide commentary in the README on how to leverage the Unix domain socket using the FileWriter, in case people want that.
* We should provide commentary in the README on how to pull in and engage the runtime metrics library.
* We should call out that the HttpClient library was removed as part of the migration instructions, just in case external teams were using it.
* I like having the migration notes in the README as a way of pointing users to a single location that can answer most of their questions.
* One thing I am a bit confused about is that I do not see a test which validates that complex Ids are translated into the spectatord line protocol. The way the meters are composed, it relies on the Id type to create the correct string, but I do not see a corresponding test for this in the Id tests. It should be something like the following, in spectator-py:

    https://github.com/Netflix/spectator-py/blob/main/tests/test_id.py#L68-L72

    This helps us ensure that the protocol line are constructed correctly with complex Ids.
* Otherwise, all of the changes look good, and are as expected, due to following the previous PRs.

File: spectator/logger/logger.go [hunk 1]
```diff
@@ -1,8 +1,8 @@
 package logger
 
 import (
-	"log"
-	"os"
+	"fmt"
+	"log/slog"
```
Comment (copperlight) [inline spectator/logger/logger.go:5]: Eh, what is the reasoning behind this change?

File: spectator/writer/udp_writer_test.go [hunk 2]
```diff
@@ -103,7 +103,7 @@ func TestUdpWriter_Write(t *testing.T) {
 }
 
 func TestConcurrentWrites(t *testing.T) {
-	messagesPerThread := 2500
+	messagesPerThread := 1000
```
Comment (copperlight) [inline spectator/writer/udp_writer_test.go:106]: We should maybe leave a comment here about what we think the actual top-end looks like, so we know what it is, even though we downshift to improve CI testing reliability.

File: README.md [hunk 3]
```diff
@@ -101,3 +102,86 @@ func main() {
 Logging is implemented with the standard Golang [slog package](https://pkg.go.dev/log/slog). The logger
 defines interfaces for [Debugf, Infof, and Errorf](./spectator/logger/logger.go). There are useful messages
 implemented at the Debug level which can help diagnose the metric publishing workflow.
+
+---
+
+## Migrating from 0.2.X to 0.3.X
```
Comment (copperlight) [inline README.md:108]: I think we are going to go for it and make this the 1.0.X version, because it fulfills the vision we have.

Python did not go full 1.0.X yet, because some legacy API cruft remains. We remove that and make the Meter handling `withId` variations the same as here, and I think we can then release that as a 1.0.X.

**Review by copperlight — APPROVED** [review]
Perfect! I love what you have done with the README rewrite!

Only question I have is seeking to understand the `log`->`slog` change a little better.

File: spectator/logger/logger.go (hunk 1 above)
Comment (jivimberg) [reply spectator/logger/logger.go:5]: The default logger doesn't support levels, so all 3 levels were always shown when using the default logger. The only way to change that behavior was by implementing a custom logger. As a consequence we had an additional mechanism to control whether the payloads were logged or not (an env variable).

Instead, I changed to `slog` which supports levels out-of-the-box. Its default log level is Info (which sounds more reasonable to me). Now to log message payloads users just change the log level to default. 

So to sum up: simplicity and principle of least surprise

Comment (copperlight) [reply spectator/logger/logger.go:5]: My understanding is that the default Go logger is always like this, and it is incumbent upon the application owner to select a more capable logging package for their application, such as Uber/zap. There is enough variation in these selections, that I think we want to keep the most minimal interface here, to ensure that it works with any logger an app owner might choose.

Comment (copperlight) [reply spectator/logger/logger.go:5]: Oh, I see it's a new addition to the standard lib as of `1.21`:

> The new log/slog package in Go 1.21 brings structured logging to the standard library.

Excellent, good find.

**Comment by copperlight** [comment]
Note that the CI failures are most likely due to the fact that the `slog` package was introduced in `1.21`.

Really, we should only be building against the latest two stable versions, which is `1.21` and `1.22`. I would drop these old versions.

File: spectator/writer/udp_writer_test.go (hunk 2 above)
Comment (jivimberg) [reply spectator/writer/udp_writer_test.go:106]: I change it down from 2500 because it was sometimes failing (though rarely).

My understanding is that it really depends on how fast the reader can consume messages on the other end so I'm not sure a test will give us an accurate figure. For example, I tried just counting the messages instead of accumulating them in an array and that changed the max threshold.

If this is a concern we should introduce a buffer.

File: README.md (hunk 3 above)
Comment (jivimberg) [reply README.md:108]: Sounds good. I'll update the README

**Comment by jivimberg** [comment]
> * We should provide commentary in the README on how to leverage the Unix domain socket using the FileWriter, in case people want that.

Implemented as a different writer because Go os.OpenFile API doesn't work for writing to Unix Domain Sockets. 
Users can use `unix://file/to/socket` as output location to write to Unix domain sockets.

File: spectator/writer/udp_writer_test.go (hunk 2 above)
Comment (copperlight) [reply spectator/writer/udp_writer_test.go:106]: Interesting observation. I can see how that makes sense. Not too worried about introducing a buffer just yet - we are looking for a basic t-shirt sizing style of the performance for now.

**Review by copperlight — APPROVED** [review]
Nice work! This feels fairly complete to me.

File: README.md [hunk 4]
```diff
@@ -102,39 +99,90 @@ func main() {
...
+  troubleshoot.
+- `spectator.Registry` no longer keep track of the Meters it creates. This means that you can't get a list of all Meters
+  from the Registry. If you need to keep track of Meters, you can do so in your application code.
+- `Percentile*` meters no longer support defining min/max values.
+
+### Migration steps
 
-[Issue#49366]: https://github.com/golang/go/issues/49366
-[PR#59]: https://github.com/Netflix/spectator-go/pull/59
+1. Make sure you're not relying on any of the [removed functionality](#removed).
```
Comment (copperlight) [inline README.md:180]: We need to add a step to mention that any dependencies on the internal configuration library should be removed.

Comment (jivimberg) [reply README.md:180]: What configuration library are you referring to?
